Branch-SDK/src/main/java/io/branch/referral/BranchLinkData.java

```diff
@@ -373,7 +368,14 @@ public boolean equals(Object obj) {
         if (tags == null) {
             if (other.tags != null)
                 return false;
-        } else if (!tags.toString().equals(other.tags.toString()))
+        } else if (!tags.toString().equals(other.tags != null ? other.tags.toString() : null))
+            return false;
+
+        // todo JSONObject's order of keys is not consistent, so this equality can lead to false positives
+        if (params == null) {
+            if (other.params != null)
+                return false;
+        } else if (!params.toString().equals(other.params != null ? other.params.toString() : null))
```

Comment:
This equality never worked before (see the comment why), now it will work most of the time but still not always. It's the best we can do though without a serialization library or a custom, recursive JSONObject equals method which would be easy to mess up and I doubt it would add much value given that this never worked and no one noticed.

Reply:
If that check didn't work before, could we just remove it?

Reply:
So `BranchLinkData.java` is used as a key to cache urls that the client creates, if we removed the equality of `params`, then url link builders with different `params` would yield the same url each time which is worse I think.

Reply:
We use "stringified" json as keys in a map to track impression on the disco SDK, it's most likely going to be changed eventually because it isn't stable but so far it seems to work reasonably well, so I figured it's a good enough improvement *for now*, I do agree that eventually it should be fixed properly.